### PR TITLE
Port forwarding in qemu runner

### PR DESCRIPTION
**Pull request by fntlnz**
I added the `-publish` option to the `qemu` runner for the linuxkit cmd

**Usage**

```bash
$ linuxkit run qemu -publish 8080:80 -publish 8081:80 example
```

or even

```bash
linuxkit run qemu -publish 1111:1111/udp -publish 8081:80/tcp example
```

**Help output**
```
Please specify the prefix to the image to boot
USAGE: linuxkit run qemu [options] prefix

'prefix' specifies the path to the VM image.

Options:
  -arch string
        Type of architecture to use, e.g. x86_64, aarch64 (default "x86_64")
  -cpus string
        Number of CPUs (default "1")
  -diskpath string
        Path to disk image to use
  -disksize string
        Size of disk to create, only created if it doesn't exist
  -fwpath string
        Path to OVMF firmware for UEFI boot (default "/usr/share/ovmf/bios.bin")
  -gui
        Set qemu to use video output instead of stdio
  -iso
        Set Legacy BIOS boot from 'prefix'.iso
  -kernel
        Set boot using 'prefix'-bzImage/-initrd/-cmdline (default true)
  -mem string
        Amount of memory in MB (default "1024")
  -publish value
        Publish a vm's port(s) to the host (default [])
  -uefi
        Set UEFI boot from 'prefix'-efi.iso
```

## Related work
This work is related to #1698 - this PR addresses only half the issue.
This is just port forwarding other part would be to do linux briding in order to have a veth pair within the guest vm and being able to access it directly.

## Moar

This is just some initial implementation of my idea, for example I put all the custom flags and build functions in the same `run_qemu` file because didn't know if there's a better place, any suggestion is warmly welcomed :angel: 

## Also

I don't know **how much** I'm breaking, there's any integration test suite I'm not aware of?

### CODE REVIEW

**Comment by fntlnz** [comment]
ping  @thebsdbox @dave-tucker @justincormack

File: src/cmd/linuxkit/run_qemu.go [hunk 1]
```diff
@@ -15,21 +17,22 @@ const QemuImg = "linuxkit/qemu:17f052263d63c8a2b641ad91c589edcbb8a18c82"
...
-	DiskSize      string
-	FWPath        string
-	Arch          string
-	CPUs          string
-	Memory        string
-	KVM           bool
-	Containerized bool
-	QemuBinPath   string
-	QemuImgPath   string
+	Prefix         string
```
Comment (rneugeba) [inline src/cmd/linuxkit/run_qemu.go]: what changed here?

File: src/cmd/linuxkit/run_qemu.go [hunk 2]
```diff
@@ -283,3 +295,77 @@ func buildPath(prefix string, postfix string) string {
 	}
 	return path
 }
+
+type multipleFlag []string
```
Comment (dave-tucker) [inline src/cmd/linuxkit/run_qemu.go]: I'd be tempted to move this to `util.go`

File: src/cmd/linuxkit/run_qemu.go [hunk 3]
```diff
@@ -283,3 +295,77 @@ func buildPath(prefix string, postfix string) string {
...
+
+type publishedPorts struct {
+	guest    int
+	host     int
+	protocol string
+}
+
+func (f *multipleFlag) String() string {
```
Comment (dave-tucker) [inline src/cmd/linuxkit/run_qemu.go]: And these functions too...

**Comment by rneugeba** [comment]
I don't think we will standardise on one set of arguments across all backends. They are just too different. It's good to have a "common" set, though and if more than one backend supports a particular feature than obviously they should use the same option name...

It would be good if the commit message contained the example or even the help text. The PR text is not easily available via the git command line...

File: src/cmd/linuxkit/run_qemu.go (hunk 3 above)
Comment (fntlnz) [inline src/cmd/linuxkit/run_qemu.go]: Now it works in a container :tada: , haven't noticed that there was a way to run it in a container, could be worth to make it configurable via a flag instead than just checking if qemu is installed?

Example use case: for this pr not having that flag required me to hardcode `true` for containerized

**Comment by fntlnz** [comment]
@rneugeba agree for the backends, opened  an issue regarding this #1732 - also because things are diverging very quickly

I added the message to the pr too :)

**Comment by fntlnz** [comment]
There's a logic issue with container forwardings, please don't merge this yet

**Comment by justincormack** [comment]
You will also need to rebase to fix CI I think

**Comment by fntlnz** [comment]
@justincormack fixed, rebased and pushed - I'm wondering about writing some integration tests, wdyt?

**Comment by justincormack** [comment]
Tests are always good!

cc @dave-tucker

**Comment by dave-tucker** [comment]
yes we should probably have some integration tests but i think we should wait until we have a proper framework for running them. Overall this LGTM!

**Comment by dave-tucker** [comment]
Note: This will probably need a rebase to pick up the latest iteration of CI changes

**Comment by justincormack** [comment]
yes, rebase should hopefully fix CI...

**Comment by fntlnz** [comment]
@dave-tucker rebased :rabbit:  CI Fixed too!

@justincormack I see that something's happening around testing like here: https://github.com/linuxkit/linuxkit/tree/master/test/cases

**Comment by dave-tucker** [comment]
@fntlnz yes indeed there is!
I've got a PR brewing with a proper test suite.
See: https://github.com/dave-tucker/linuxkit/tree/rtf-cases/test/cases for a sneak preview

Hopefully we can add a nice integration test in`test/cases/010_platforms/000_qemu/` soon, but I don't think this should prevent merging this PR.

**Pull request merged**